### DISCUSSION TITLE
issue: Possible race condition when resetting a form with a form array

### DISCUSSION THREAD

**Discussion by Mario-Eis** [react-hook-form/react-hook-form#8996] [Q&A]
### Version Number

7.34.2

### Codesandbox/Expo snack

I was not able to do this

### Steps to reproduce

I am sitting on an error that drives me nuts. Basically I created a form, that receives data from the backend and presents it for editing. The user can click "next" to navigate to a new router path and load another entry. The data consists of an array of objects with fields of information.

But randomly it fails to reset the form correctly. The arrays (the old state and the new default values) simply gets merged together.

At first I thought the issues would be related to https://github.com/react-hook-form/react-hook-form/issues/3378. But I guess that is not the case. The effect order should be correct.

I tried really hard to create a sandbox. But I failed :( I did see some very similar effects with very long lists and material ui components in the sandbox. But they were very brief. Not as consistent and for me not reliable enough to say they are related to this issue.

But I did a lot of tests. And one thing I found out: The more complex the page gets the more often the reset fails that way. Also things like the dev tools are noticeable here. Maybe that's why I was not able to create a sandbox. 
Another thing I could figure out: When the useEffect for the form reset is changed to useLayoutEffect, the issue stops. That is why I think it could be a race condition.

Here is the code I used:

```
export const WorkflowForm: FunctionComponent<WithClassname> = ({className}) => {
    const componentId = useComponentId();
    const routeParams = useParams<OrderManagementRouteParams>();

    const {data: workflow} = useReadWorkflow(routeParams.orderItemId);

    const initialFormData = useMemo(() => ({workflow}), [workflow]);

    const methods = useForm<FormData>({
        defaultValues: initialFormData
    });

    console.log(initialFormData)

    const {fields} = useFieldArray({control: methods.control, name: "workflow"});

    useEffect(() => {
        methods.reset(initialFormData)
    }, [initialFormData, methods]);

    const onSubmit = methods.handleSubmit(() => null);

    return <FormProvider {...methods}>
        <form className={className} onSubmit={onSubmit}>
            {
                fields?.map((workflowStep: WorkflowStep, index: number) => {
                    return <div key={workflowStep.id || `${componentId}.new.${index}`}>
                        <Controller name={`workflow[${index}].data.info`} render={({field}) => <TextField {...field} ref={null} inputRef={field.ref} value={field.value || ""}/>}/>
                    </div>;
                })
            }
        </form>
    </FormProvider>;
};
```

And here is a log output (see code above). It shows the old default values with an array of five entries and the new one with just one entry. The form state is rendered as input fields. There are five instead of one where the first one is the single new entry.

![2022-09-06 20_08_29-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188709955-637eb5ec-326c-4b04-8bbc-e40911505b06.png)

### Expected behaviour

The form should reset to the new data.

### What browsers are you seeing the problem on?

Chrome

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
Does the issue persist if you remove this line?
> key={workflowStep.id || `${componentId}.new.${index}`}

**Comment by Mario-Eis** [comment]
Yes unfortunately it does. I also was thinking in this direction. But I can cross out a lot of react or memo related issues as well. Because the react hook form dev tools show the internal state as the merged array. And "fields" also contains the merged array.
The only thing I could do is useLayoutEffect for the reset. This works around the issue.

**Comment by Mario-Eis** [comment]
I am also working on a sandbox. But I was able to produce this strange state in a stable manor only once. ~~When in the sandbox more often I am able to get a state where it merges the arrays as well, but fixes its length in a blink of an eye later.~~ UPDATE: This is not correct. It was the double useEffect from strict mode I was seeing.

**Comment by Mario-Eis** [comment]
I created three more logs with the dev tools open. First is the one with the initially loaded array. And the correct output.

![2022-09-06 20_54_08-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188716567-719d36c3-1e73-49a7-b0b2-b4d795938087.png)

Next the merged, faulty state after loading data async:
![2022-09-06 20_54_32-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188716667-9ef7ad82-68b1-4bf6-b288-67285fd24d98.png)

And here is the correct second state after a page refresh:
![2022-09-06 20_55_33-FIVA - DEV - Auftragsverwaltung - Opera](https://user-images.githubusercontent.com/17924929/188716802-819614f5-db9f-4d3d-918c-b616df7cb97c.png)

The console logs show what is passed as default values.

**Comment by bluebill1049** [comment]
> Another thing I could figure out: When the useEffect for the form reset is changed to useLayoutEffect, the issue stops. That is why I think it could be a race condition.

we need a CSB/unit tests to reproduce the issue, it's rather difficult to isolate the problem otherwise.

**Comment by Mario-Eis** [comment]
Got it! I figured out, that our navigation loads the id of the next item before it actually navigates. So what is going on is an async request, a navigation and another async request. I pined it down to the async request triggering another async request. I don't know what exactly is going on with the form state, but this is what somehow destroys it. We switched over from formik to react-hook-form and did not think this could be an issue. So it was a tough one...it's 1am already :D

Here is a sandbox link: https://codesandbox.io/s/react-hook-form-conditional-forked-8bkns9?file=/src/index.js

It should be enough to klick next -> previous -> next. And you get this strange "merged state". You can also try to use useLayoutEffect and the issue will stop.

**Comment by bluebill1049** [comment]
you are not following the doc on the key implementation for field array

```diff
<ul>
  {fields?.map((item, index) => {
    return (
-      <li key={index}>
+      <li key={item.id}>
        <TextInputField name={`receivedData.${index}.first_name`} />
        <TextInputField name={`receivedData.${index}.last_name`} />
        <TextInputField name={`receivedData.${index}.email`} />
      </li>
    );
  })}
</ul>
```

react hook form shares different philosophy and design pattern than other existing form libraries (controlled), so I would pay attention to those details in the doc. I am sure the switching will create a much smoother experience and accessibility for your customer on low-end devices/computers.

**Comment by Mario-Eis** [comment]
ha...
You are completely right with the sandbox. But I did the same thing in the products code before (got several hints already). And it made absolutely no difference. Here is the code:

```
const {fields} = useFieldArray({control: methods.control, name: "workflow", keyName: "key"});

... 

fields?.map((field: WorkflowStep & { key: string }, index: number) => {
    return <div key={field.key}>
          <TextInputField name={`workflow[${index}].data.info`}/>
     </div>;
 })

...

let TextInputField: FunctionComponent<TextFieldProps> = ({name, ...props}) => {
    const {control} = useFormContext();
    const value = useWatch({control, name});
    const {field: {ref, ...field}} = useController({name, control, defaultValue: value});

    return <TextField {...props} {...field} inputRef={ref}/>;
};
```

**Comment by Mario-Eis** [comment]
I slightly modified the sandbox. Click next and previous in rapid succession. After 2-3 cycles the state ends up "merged" again (you possibly have to try a bit), although the key is set correctly now. With the small change, that it (sometimes) "repairs" itself when clicking a button again. That is more like my prod code is behaving.

https://codesandbox.io/s/react-hook-form-conditional-forked-8bkns9?file=/src/index.js

**Reply by bluebill1049** [reply]
works fine for me without those setTimeout hacks: https://codesandbox.io/s/react-hook-form-conditional-forked-9momn9

**Reply by Mario-Eis** [reply]
But they are the key. Our application is loading the next entry id before loading the next entry itself. Because navigation is done with the browser history. I simulated this by adding setTimeout. The flow is: click "next" -> load next id -> push browser history with the new id -> receive new route params in the form component -> load workflow data -> show form.

**Comment by Mario-Eis** [comment]
Update: I was also able to work around the issue by dropping out the whole component when fetching data.

`!isLoadingWorkflow && <WorkflowFormComponent className={className} workflow={workflow} workSteps={workSteps}/>;`

But this is more like a hard reset hack. Completely dropping the whole form without using the reset API. It also causes bad flickering. And I don't think this is the idea of resetting the form...

**Comment by Mario-Eis** [comment]
Another update: I found another clue or workaround. Watching the field array instead of using useFieldArray's return value, also seems to get somehow rid of the issue.

```
    const {append: addWorkflowStep, remove: removeWorkflowStep, move: moveWorkflowStep} = useFieldArray({
        control: methods.control,
        name: "workflow"
    });

    const fields = methods.watch("workflow");
```

BUT: With this method, the real ID of the business object is used as the key value in the array mapping. And that's not how its intended I guess. So I don't know the implications of that.

A possible stabilization would maybe be 

```
    const {fields: realFields, append: addWorkflowStep, remove: removeWorkflowStep, move: moveWorkflowStep} = useFieldArray({
        control: methods.control,
        name: "workflow"
    });

    const fields = methods.watch("workflow")?.map((field, index) => {
        field.id = realFields[index].id;
        return field;
    });
```
This causes a short flickering of the old data in the new (reset) form. I think there is one render cycle using the previous IDs. But also works at the end of the day. More or less.

**Comment by leapful** [comment]
@Mario-Eis, @bluebill1049  : 

The prop "fields" from "useFieldArray" will be updated after resetting data but then "useQuery" of "react-query" is executed asynchronously in case rapid succession therefore updated remote data from API is not correct for next rendering of "react-hook-form".

You can perform to "useForm reset()" with "useLayoutEffect" in order to keep react-hook-form state correct at the time DOM elements are not mounted yet, then perform "useFieldArray replace()" with "useEffect" to replace correctly the "useFieldArray fields" data.

For example:
```
const { reset, control } = useForm();

const { fields, replace }  = useFieldArray({ name: 'receivedData',  control })

useLayoutEffect(() => {
  reset(remoteData);
}, [remoteData, reset]);

useEffect(() => {
   replace(remoteData.receivedData);
}, [remoteData, replace]);
```

Working demo: https://codesandbox.io/s/react-hook-form-conditional-forked-diyfg8?file=/src/index.js

**Reply by Mario-Eis** [reply]
Wouldn't replacing data leaving the form dirty? But you brought another idea to my mind. I could try to swap the query with the useForm hook. Thus also swap their internal side effect order. In that scenario it would not be possible to pass default values to the useForm hook. BUT to tigger an initial reset side effect afterwards. I will try it later!

**Reply by leapful** [reply]
@Mario-Eis : The dirty state is only updated if default values are not equals with current values. At phase "useLayoutEffect", "reset" already set values to new data then the same data is put into "replace" therefore no dirty field exists at this time.

**Comment by AAS97** [comment]
Hi there !
What is the status on this discussion ? Any solution in the making ?

**Reply by Mario-Eis** [reply]
For my situation I am still going with the useLayoutEffect workaround **for now**.
